core/logger/audit/audit_logger.go

```diff
@@ -0,0 +1,231 @@
+	}
+
+	loggingChannel := make(chan WrappedAuditLog, bufferCapacity)
+
+	// Finally, create new auditLogger with parameters
+	auditLogger := AuditLoggerService{
+		logger:          logger.Helper(1),
+		enabled:         true,
+		serviceURL:      config.forwardToUrl,
+		serviceHeaders:  config.headers,
+		jsonWrapperKey:  config.jsonWrapperKey,
+		environmentName: config.environment,
+		hostname:        hostname,
+		localIP:         getLocalIP(),
+		loggingChannel:  loggingChannel,
+	}
+
+	// Start our go routine that will receive logs and send them out to the
+	// configured service.
+	go auditLogger.auditLoggerRoutine()
+
+	return &auditLogger, nil
+}
+
+// / Entrypoint for new audit logs. This buffers all logs that come in they will
+// / sent out by the goroutine that was started when the AuditLoggerService was
+// / created. If this service was not enabled, this immeidately returns.
+// /
+// / This function never blocks.
+func (l *AuditLoggerService) Audit(ctx context.Context, eventID EventID, data Data) {
+	if !l.enabled {
+		return
+	}
+
+	wrappedLog := WrappedAuditLog{
+		eventID: eventID,
+		data:    data,
+	}
+
+	select {
+	case l.loggingChannel <- wrappedLog:
+	default:
+		if l.loggingChannel == nil {
+			l.logger.Errorw("Could not send log to audit subsystem because it has gone away!")
+		} else {
+			l.logger.Errorw("Audit log buffer is full. Dropping log with eventID: %s", eventID)
+		}
+	}
+}
+
+// / Entrypoint for our log handling goroutine. This waits on the channel and sends out
+// / logs as they come in.
+// /
+// / This function calls postLogToLogService which blocks.
+func (l *AuditLoggerService) auditLoggerRoutine() {
+	for event := range l.loggingChannel {
+		l.postLogToLogService(event.eventID, event.data)
+	}
+
+	l.logger.Errorw("Audit logger is shut down. Will not send requested audit log")
+}
+
+// / Takes an EventID and associated data and sends it to the configured logging
+// / endpoint. This function blocks on the send by timesout after a period of
+// / several seconds. This helps us prevent getting stuck on a single log
+// / due to transient network errors.
+// /
+// / This function blocks when called.
+func (l *AuditLoggerService) postLogToLogService(eventID EventID, data Data) {
+	// Audit log JSON data
+	logItem := map[string]interface{}{
+		"eventID":  eventID,
+		"hostname": l.hostname,
+		"localIP":  l.localIP,
+		"env":      l.environmentName,
+		"data":     data,
+	}
+
+	// Optionally wrap audit log data into JSON object to help dynamically structure for an HTTP log service call
+	if l.jsonWrapperKey != "" {
+		logItem = map[string]interface{}{l.jsonWrapperKey: logItem}
+	}
+
+	serializedLog, err := json.Marshal(logItem)
+	if err != nil {
+		l.logger.Errorw("Unable to serialize wrapped audit log item to JSON", "err", err, "logItem", logItem)
+		return
+	}
+
+	// Send up to HEC log collector
+	httpClient := &http.Client{Timeout: time.Second * 10}
```

Comment:
This should probably be set on the AuditLoggerService and not re-used each time.

You may also consider passing it into the constructor, this would allow you to easily mock it in a test.